src/pdf.ts

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  Widget
+} from '@phosphor/widgets';
+
+import {
+  IRenderMime
+} from '@jupyterlab/rendermime-interfaces';
+
+declare var require: any;
+let PDF = require('pdfjs-dist');
+PDF = require('pdfjs-dist/web/pdf_viewer');
+
+import '../style/index.css';
+import 'pdfjs-dist/web/pdf_viewer.css';
+
+/**
+ * The MIME type for PDF.
+ */
+export
+const MIME_TYPE = 'application/pdf';
+
+export
+const PDF_CLASS = 'pdfViewer';
+
+export
+const PDF_CONTAINER_CLASS = 'jp-PDFJSContainer';
+
+/**
+ * A class for rendering a PDF document.
+ */
+export
+class RenderedPDF extends Widget implements IRenderMime.IRenderer {
+  constructor() {
+    super({ node: Private.createNode() });
+    PDF.PDFJS.disableWorker = true;
+    this._pdfViewer = new PDF.PDFJS.PDFViewer({
+        container: this.node,
+    });
+  }
+
+  /**
+   * Render PDF into this widget's node.
+   */
+  renderModel(model: IRenderMime.IMimeModel): Promise<void> {
+    return new Promise<void>(resolve => {
+      let data = model.data[MIME_TYPE] as string;
+      // If there is no data, do nothing.
+      if (!data) {
+        resolve (void 0);
+      }
+      const blob = Private.b64toBlob(data, MIME_TYPE);
+
+      let oldUrl = this._objectUrl;
+      this._objectUrl = URL.createObjectURL(blob);
+
+      //Try to keep the scroll position.
+      const scrollTop = this.node.scrollTop;
+
+      PDF.PDFJS.getDocument(this._objectUrl).then((pdfDocument: any) => {
+        this._pdfViewer.setDocument(pdfDocument);
+        this._pdfViewer.pagesPromise.then(() => {
+          this.node.scrollTop = scrollTop;
+          resolve(void 0);
+        });
+      });
+
+      // Release reference to any previous object url.
+      if (oldUrl) {
+        try {
+          URL.revokeObjectURL(oldUrl);
+        } catch (error) { /* no-op */ }
+      }
+    });
+  }
+
+  /**
+   * Dispose of the resources held by the pdf widget.
+   */
+  dispose() {
+    try {
+      URL.revokeObjectURL(this._objectUrl);
+    } catch (error) { /* no-op */ }
+    super.dispose();
```

Comment:
The dispose, or the object URL stuff?

Reply:
The `dispose()` method gives the object a chance to free any resources that it holds. So here, for instance, we use the opportunity to clear the objectURL that has been set up. Then we call `super.dispose()` on the base class in case it wants to clean up any of its own resources.